File: leetcode/maximum_product_subarray/maximum_product_subarray.java
package maximum_product_subarray;

class Solution {
    public int maxProduct(int[] nums) {
        int currMax = nums[0], currMin = nums[0], ans = nums[0];
        
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] < 0) {
                int temp = currMax;
                currMax = currMin;
                currMin = temp;
            }
            
            currMax = Math.max(nums[i], currMax * nums[i]);
            currMin = Math.min(nums[i], currMin * nums[i]);
            
            ans = Math.max(ans, currMax);
        }
        
        return ans;
    }
}

File: leetcode/maximum_product_subarray/maximum_product_subarray.cpp
/*
Time Complexity:  O(n)
Space Complexity: O(1)
*/

#include <vector>

using std::vector;
using std::max;
using std::min;
using std::swap;

class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int currMax = nums[0], currMin = nums[0], ans = nums[0];

        for (int i = 1; i < nums.size(); i++) {
            if (nums[i] < 0) {
                swap(currMax, currMin);
            }

            currMax = max(nums[i], currMax * nums[i]);
            currMin = min(nums[i], currMin * nums[i]);

            ans = max(ans, currMax);
        }

        return ans;
    }
};